nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+		}
+	}
+	e.Status = stResolved
+	e.Flags = flags
+	e.putTime = now
+	e.validUntil = now.Add(time.Second * time.Duration(minTtl))
+	e.reply = *reply
+
+	if shouldNotify {
+		close(e.waitChan) // notify all the waiters by closing the channel
+	}
+}
+
+// wait until a valid reply is set in the cache
+func (e *cacheEntry) waitReply(request *dns.Msg, timeout time.Duration, now time.Time) (*dns.Msg, error) {
+	if e.Status == stResolved {
+		return e.getReply(request, now)
+	}
+
+	if timeout > 0 {
+		select {
+		case <-e.waitChan:
+			return e.getReply(request, now)
+		case <-time.After(time.Second * timeout):
+			return nil, errTimeout
+		}
+	}
+
+	return nil, errCouldNotResolve
+}
+
+func (e *cacheEntry) close() {
+	if e.Status == stPending {
+		close(e.waitChan)
+	}
+}
+
+// entriesSlice is used for sorting entries
+type cacheEntriesSlice []*cacheEntry
+
+func (p cacheEntriesSlice) Len() int           { return len(p) }
+func (p cacheEntriesSlice) Less(i, j int) bool { return p[i].validUntil.Before(p[j].validUntil) }
+func (p cacheEntriesSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+type cacheKey struct {
+	question dns.Question
+	protocol dnsProtocol
+}
+
+type entries map[cacheKey]*cacheEntry
+
+// Cache is a thread-safe fixed capacity LRU cache.
+type Cache struct {
+	Capacity int
+
+	entries entries
+	lock    sync.RWMutex
+}
+
+// NewCache creates a cache of the given capacity
+func NewCache(capacity int) (*Cache, error) {
+	if capacity <= 0 {
+		return nil, errInvalidCapacity
+	}
+	c := &Cache{
+		Capacity: capacity,
+		entries:  make(entries, capacity),
+	}
+	return c, nil
+}
+
+// Clear removes all the entries in the cache
+func (c *Cache) Clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entries = make(entries, c.Capacity)
+}
+
+// Purge removes the old elements in the cache
+func (c *Cache) Purge(now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.removeOldest(c.Capacity, now)
+}
+
+// Add adds a reply to the cache.
+func (c *Cache) Put(request *dns.Msg, reply *dns.Msg, proto dnsProtocol, flags uint8, now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if ent, found := c.entries[key]; found {
+		ent.setReply(reply, flags, now)
+	} else {
+		// If we will add a new item and the capacity has been exceeded, make some room...
+		if len(c.entries) >= c.Capacity {
+			c.removeOldest(1, now)
+		}
+		c.entries[key] = newCacheEntry(&question, reply, stResolved, now)
+	}
+}
+
+// Look up for a question's reply from the cache.
+// If no reply is stored in the cache, it returns a `nil` reply and no error. The caller can then `Wait()`
+// for another goroutine `Put`ing a reply in the cache.
+func (c *Cache) Get(request *dns.Msg, proto dnsProtocol, now time.Time) (reply *dns.Msg, err error) {
```

Comment:
I think this will be the typical use case with truncated responses:

_1)_ the client sends a non-local query to WeaveDNS with UDP
_2)_ WeaveDNS sends the same request to the fallback server with UDP
_3)_ the fallback server realizes the response will not fit in a UDP packet, so it fills the packet with as much data as it can but it marks the packet as _truncated_ so the client will know some info is missing
_4)_ the fallback server sends this truncated packet to WeaveDNS
_5)_ WeaveDNS caches the _"UDP+truncated"_ response
_6)_ WeaveDNS sends it back to the client
_7)_ the client sees the _truncated_flags and realizes there is some info missing
_8)_ the client retries the query, but this time with TCP
_9)_ WeaveDNS proxies the query to the fallback server, this time with TCP
_10)_ the fallback server sends all the info to WeaveDNS through the TCP connection
_11)_ WeaveDNS caches the _"TCP"_ response
_12)_ WeaveDNS sends the info to the client
_13)_ the client is done

So we will end up having the "UDP+truncated" response in the cache as well as the "TCP" response. Both responses will be immediately returned to clients the next time someone asks about the same query, so we will skip _2_, _3_, _4_, _5_, _9_, _10_ and _11_.